test/test_misc_utils.py

```diff
@@ -1156,3 +1156,115 @@ def test_file_contents():
+    assert counter() == 0
+    assert counter() == 1
+    assert counter() == 2
+
+
+class TestCachedField:
+
+    DEFAULT_TIMEOUT = 600
+
+    def test_cached_field_basic(self):
+        def simple_update_function():
+            return random.choice(range(10000))
+
+        field = CachedField('simple1', update_function=simple_update_function)
+        assert field.value is not None
+        current = field.get()
+        assert current == field.value
+        assert field.get_updated() != current
+        assert field.timeout == self.DEFAULT_TIMEOUT
+        field.set_timeout(30)
+        assert field.timeout == 30
+
+    def test_cached_field_mocked(self):
+        dt = ControlledTime()
+
+        with mock.patch.object(datetime_module, "datetime", dt):
+            field = CachedField('simple1', update_function=make_counter())
+
+            assert field.value is not None
+
+            # Get a value, which should not be changing over short periods of time.
+            val1 = field.get()
+
+            assert field.value == val1
+            assert field.get() == val1
+
+            assert field.value == val1
+            assert field.get() == val1
+
+            # Forcing an update even though not much time has passed. Field value should be updated.
+            val2 = field.get_updated()
+
+            assert val2 != val1
+            assert field.value != val1
+            assert field.value == val2
+
+            # Immediately recheck value, but accepting cache value. Field value should be unchanged.
+            val3 = field.get()
+
+            assert val3 == val2
+            assert field.value == val2
+            assert field.value == val3
+
+            # Wait a while, but not enough to trigger cache update. Field value should be unchanged.
+            dt.sleep(self.DEFAULT_TIMEOUT / 2)
+            val4 = field.get()
+
+            assert val2 == val4
+            assert val2 == field.value
+            assert val4 == field.value
+
+            # Wait a bit longer, enough to trigger cache update. Field value should be updated.
+
+            dt.sleep(self.DEFAULT_TIMEOUT / 2)  # This should push us into the cache refill time
+            val5 = field.get()
+
+            assert val2 != val5
+            assert field.value != val2
+            assert field.value == val5
+
+            assert field.get() == val5  # This is the new stable value until next cache timeout
+            assert field.get() == val5
+            assert field.get() == val5
+            assert field.get() == val5
+
+            dt.sleep(self.DEFAULT_TIMEOUT)  # Fast forward to where we're going to refill again
+            assert field.get() != val5
+
+    def test_cached_field_timeout(self):
+        field = CachedField('simple1', update_function=make_counter())
+
+        assert field.timeout == self.DEFAULT_TIMEOUT
+        field.set_timeout(30)
+        assert field.timeout == 30
+
+
+@pytest.mark.parametrize('input, expected', [
```

Comment:
Probably it should have something like `XMLContainer` and `X_M_L_Container` to emphasize the weird thing that is likely to happen there even if it follows from the ABCD example. Mostly.